### PR TITLE
[SPARK-11469] [SQL] Allow users to define nondeterministic udfs.

### PR DESCRIPTION
**Pull request by yhuai**
This is the first task (https://issues.apache.org/jira/browse/SPARK-11469) of https://issues.apache.org/jira/browse/SPARK-11438

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 1]
```diff
@@ -191,4 +193,86 @@ class UDFSuite extends QueryTest with SharedSQLContext {
...
+    }
+
+    {
+      val df = sql("SELECT 1 as a")
+        .select(col("a"), nondeterministicUDF(col("a")).as("b"))
+        .select(col("a"), col("b"), nondeterministicUDF(col("b")).as("c"))
+      checkNumUDFs(df, 2)
+      checkAnswer(df, Row(1, 2, 3))
+    }
+  }
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: @mengxr Here are two examples to define nondeterministic UDFs.

**Comment by yhuai** [comment]
cc @marmbrus

**Comment by yhuai** [comment]
test this please

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 2]
```diff
@@ -191,4 +193,86 @@ class UDFSuite extends QueryTest with SharedSQLContext {
...
   }
+
+  private def checkNumUDFs(df: DataFrame, expectedNumUDFs: Int): Unit = {
+    val udfs = df.queryExecution.optimizedPlan.collect {
+      case p: logical.Project => p.projectList.flatMap {
+        case e => e.collect {
+          case udf: ScalaUDF => udf
+        }
+      }
+    }.flatMap(functions => functions)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: We can just call `faltten`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 3]
```diff
@@ -30,13 +30,16 @@ case class ScalaUDF(
...
-    inputTypes: Seq[DataType] = Nil)
+    inputTypes: Seq[DataType] = Nil,
+    isDeterministic: Boolean = true)
   extends Expression with ImplicitCastInputTypes with CodegenFallback {
 
   override def nullable: Boolean = true
 
   override def toString: String = s"UDF(${children.mkString(",")})"
 
+  override def deterministic: Boolean = isDeterministic && children.forall(_.deterministic)
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: We probably need to override the `foldable` as well. Something like:
```scala
override def foldable: Boolean = deterministic && children.forall(_.foldable)
```
Or change it in `Expression` directly?

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 4]
```diff
@@ -191,4 +193,86 @@ class UDFSuite extends QueryTest with SharedSQLContext {
...
+
+    val deterministicUDF = sqlContext.udf.register("plusOne1", (x: Int) => x + 1)
+    val nondeterministicUDF = deterministicUDF.nonDeterministic
+    sqlContext.udf.register("plusOne2", nondeterministicUDF)
+
+    {
+      val df = sql("SELECT 1 as a")
+        .select(col("a"), deterministicUDF(col("a")).as("b"))
+        .select(col("a"), col("b"), deterministicUDF(col("b")).as("c"))
+      checkNumUDFs(df, 3)
```
Comment (chenghao-intel) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: The optimization rule `ConstantFolding` probably doesn't work as still the UDF number is 3, but should be 0, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: I think we can't fold any nondeterministic expressions, we should make it a common behaviour of `Expression`

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: This does seem reasonable.

File: project/MimaExcludes.scala [hunk 5]
```diff
@@ -112,6 +112,53 @@ object MimaExcludes {
           "org.apache.spark.rdd.MapPartitionsWithPreparationRDD"),
         ProblemFilters.exclude[MissingClassProblem](
           "org.apache.spark.rdd.MapPartitionsWithPreparationRDD$")
+      ) ++ Seq(
+        ProblemFilters.exclude[MissingMethodProblem](
```
Comment (marmbrus) [inline project/MimaExcludes.scala]: I'm a little confused what is going on here?

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala [hunk 6]
```diff
@@ -86,9 +104,9 @@ class UDFRegistration private[sql] (sqlContext: SQLContext) extends Logging {
         def register[$typeTags](name: String, func: Function$x[$types]): UserDefinedFunction = {
           val dataType = ScalaReflection.schemaFor[RT].dataType
           val inputTypes = Try($inputTypes).getOrElse(Nil)
-          def builder(e: Seq[Expression]) = ScalaUDF(func, dataType, e, inputTypes)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: @marmbrus This kind of `builder`s caused the mima failure. Now, I am adding a build function in `ScalaUDF` because we need to get the value of `nondeterministic`.

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala (hunk 4 above)
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: The default value of `foldable` is false, which is why we see three expressions at here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 7]
```diff
@@ -30,13 +30,18 @@ case class ScalaUDF(
...
-    inputTypes: Seq[DataType] = Nil)
+    inputTypes: Seq[DataType] = Nil,
+    isDeterministic: Boolean = true)
   extends Expression with ImplicitCastInputTypes with CodegenFallback {
 
   override def nullable: Boolean = true
 
   override def toString: String = s"UDF(${children.mkString(",")})"
 
+  override def foldable: Boolean = deterministic && children.forall(_.foldable)
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: I override `foldable` at here. We can later take a look at how we override `foldable` in other places and see if we need to make any change.

File: project/MimaExcludes.scala (hunk 5 above)
Comment (yhuai) [inline project/MimaExcludes.scala]: https://github.com/apache/spark/pull/9393#discussion-diff-43656205 shows the reason of these mima things.

**Comment by mengxr** [comment]
@yhuai This only applies to UDFs. What if a user uses built-in expressions, e.g.:

~~~scala
val a = b + c
val d = a * a
~~~

It may be useful to write

~~~scala
val a = (b + c).nondeterministic
val d = a * a
~~~

Btw, I would prefer `udf { x => x * x }.nondeterministic` instead of `nondeterministicUdf { x => x * x }`. But it is minor.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala [hunk 8]
```diff
@@ -381,7 +381,7 @@ class ParquetIOSuite extends QueryTest with ParquetTest with SharedSQLContext {
       sqlContext.udf.register("div0", (x: Int) => x / 0)
       withTempPath { dir =>
         intercept[org.apache.spark.SparkException] {
-          sqlContext.sql("select div0(1)").write.parquet(dir.getCanonicalPath)
+          sqlContext.range(1, 2).selectExpr("div0(id) as a").write.parquet(dir.getCanonicalPath)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala]: why this change?

Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala]: It will prevent us from triggering constant folding.

Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala]: Ah, because we can do constant folding to `ScalaUDF` now. This is to make sure we will throw exception at executor side.

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
retest this please

**Comment by yhuai** [comment]
@mengxr I am going to merge it. We can change the name of the public API in a follow-up PR.
